pages/blog/internals-of-nova-part-2.md

```diff
@@ -0,0 +1,286 @@
+
+```rs
+struct ArrayHeapData {
+    /// u32 index to Vec<ObjectHeapData> or 0 as None
+    backing_object: Option<OrdinaryObject>,
+    /// u32 index to Vec<ElementsHeapData<cap>>
+    elements: ElementIndex,
+    /// u8 identifier of elements array capacity (powers of two)
+    cap: ElementCapacity,
+    /// value of this Array's length property
+    len: u32,
+    /// writable flag of this Array's length property
+    len_writable: bool,
+}
+```
+
+This is 16 bytes in total. Split into 5 homogenous slices this could be brought down to 14 bytes through the removal of two padding bytes. Combining `len_writable` into `cap` gets us to 13 bytes. But can we go lower? And should we?
+
+In the last installment I mentioned that the backing object needs to perform double-duty as a `Realm` indicator for when the backing object hasn't been created yet. If the Array knows what `Realm` created it by association, then we can drop that need. This association would be a `Realm`-specific `Heap`.
+
+If each `Realm` has its own `Heap` struct, then we do not need the backing object's memory unless a backing object has specifically been created. This means that the `backing_object` `Option` is most often `None`. Now, what is an efficient data structure that can associate an index with a another index? Yes indeed, a hashmap! In the future if Nova splits the singular `Heap` up into multiple ones then we can move to the following kind of heap structure:
+
+```rs
+struct ArrayHeapData<const N: usize> {
+    elements: [ElementIndex; N],
+    lens: [u32; N],
+    caps: [ElementCapacityWritability; N],
+}
+struct ArrayVec {
+    // Note: Invalid Rust, 'cap' cannot be referred to as const generic. This is only to show the idea.
+    ptr: *mut ArrayHeapData<cap>,
+    // Note: Use u32's as we index using u32; usize would be unnecessarily large.
+    len: u32,
+    cap: u32,
+    backing_objects: HashMap<Array, OrdinaryObject>,
+}
+```
+
+Now the common case for a JavaScript Array needs only 9 bytes of memory. Iterating over multiple Arrays to access one "part" of them only accesses those parts and none of the others. Cache line utilisation heretofore unimaginable in a JavaScript engine is attainable as easily as just writing the code. This ... this could change things.
+
+## Post scriptum: What about Objects?
+
+Let's take a quick detour to Objects as well. First thing is that we drop out the "elements" store from ordinary objects. Objects usually have named properties, only rarely do they have indexed properties. If what you want is indexed properties, then an Array is what you should use.
+
+A `Map` like in V8 is still very much necessary, though we will call it a `Shape`: Without shapes, a JavaScript engine cannot optimise prototype accesses and that then means that calling prototype methods becomes _very_ expensive indeed. That is then basically all we need, and here is what we're left with:
```

Comment:
**confusion**: What is a `Shape`? Would be nice with some more context/a basic explanation for us noobz here.

Reply:
Ah, that's a great question. I'll probably have to just link to V8's explanation on them.

Basically, a Shape describes an object's, well, shape. For V8 it includes things like the size of the object in bytes, number of inobject properties, object prototype, object property descriptors (in order), and other relevant metadata. The idea is that a JS program will often create similar objects over time, and the number of different kinds of objects in the program is limited. Hence, it makes sense to separate the shared parts of an object from the instance specific parts.

This way, at the engine level these two objects share the same shape and thus the property keys and prototype field exists only once. Only the values are saved per-instance.
```ts
const a = { x: 3, y: 4 };
const b = { x: 0, y: 5 };
```

This system also allows caching the `hasOwnProperty` property lookup result in this case:
```ts
function hasX(obj) {
  return obj.hasOwnProperty("x");
}

hasX(a);
hasX(b);
```
The `hasX` function gets called with two different objects, but those objects have the same shape so they will get the same result for `hasOwnProperty` lookup. Thus, an inline cache can be arranged:
```rs
struct InlineCache {
  shape: *mut Shape,
  result: Option<Value>,
}
```
We check if `obj.shape == inline_cache.shape` and if it is equal, then we can return `inline_cache.result` directly. If someone goes and does something evil like `Object.getPrototypeOf(a).hasOwnProperty = () => false` then a "validity cell" mechanism gets triggered that invalidates all inline caches that depend on the modified object's shape.